libs/pageserver_api/src/models.rs

```diff
@@ -979,11 +979,32 @@ impl PagestreamFeMessage {
         //
         // TODO: consider using protobuf or serde bincode for less error prone
         // serialization.
-        let msg_tag = body.read_u8()?;
+        let mut msg_tag = body.read_u8()?;
+        //
+        // Old version of protocol use commands with tags started with 0 and containing `latest` flag.
+        // New version of protocol shift command tags by 10 and pass LSN range instead of `latest` flag.
+        // Server should be able to handle both protocol version. As far as we are not passing no=w,
+        // protocol version from client to server, we make a decision based on tag range.
+        // So this code actually provides backward compatibility.
+        //
+        let horizon = if msg_tag >= 10 {
+            // new protocol
+            msg_tag -= 10; // commands tags in new protocol starts with 10
```

Comment:
Why do we need to mutate `msg_tag` here?

Reply:
It's a hack to workaround missing protocol version field. Instead msg_tag is abused, shifting all tags on 10 (we have only a few of them, so no valid values >= 10). To actually parse we need to subtract 10 back then.

Reply:
I think it would be more clear to add new commands with new tags and add handlers for them, as if they're completely new commands. After we don't need the backwards compatibility anymore, remove the old ones, leaving their tags unused.

Reply:
Unfortunately there are now no request tags in Rust at all!
There is enum ` PagestreamFeMessage`:
```
pub enum PagestreamFeMessage {
    Exists(PagestreamExistsRequest),
    Nblocks(PagestreamNblocksRequest),
    GetPage(PagestreamGetPageRequest),
    DbSize(PagestreamDbSizeRequest),
    GetSlruSegment(PagestreamGetSlruSegmentRequest),
}
```
But it is not somehow mapped to numbers. Number are hardcoded in `PagestreamFeMessage::parse`:
```
        match msg_tag {
            0 => Ok(PagestreamFeMessage::Exists(PagestreamExistsRequest {
```
and  `PagestreamFeMessage::serialize`:
```
   pub fn serialize(&self) -> Bytes {
        let mut bytes = BytesMut::new();

        match self {
            Self::Exists(req) => {
                bytes.put_u8(0);
```
 So it can be certainly refactored, but should it new part of this PR?

Reply:
We should use serde here, but it doesn't need to be done on this PR. There are already comments for using serde. What exists can be turned into bincode quite easily.

Looking at the tests, we don't have image-based testing; we only have roundtrip parsing. That is insufficient to ensure backwards-compatible changes, so these will need to be fixed at the same time.